src/backend/distributed/metadata/pg_get_object_address.c

```diff
@@ -0,0 +1,503 @@
+			break;
+		}
+
+		case OBJECT_PUBLICATION_REL:
+		{
+			objnode = (Node *) list_make2(name, linitial(args));
+			break;
+		}
+
+		case OBJECT_USER_MAPPING:
+		{
+			objnode = (Node *) list_make2(linitial(name), linitial(args));
+			break;
+		}
+
+		case OBJECT_DEFACL:
+		{
+			objnode = (Node *) lcons(linitial(args), name);
+			break;
+		}
+
+		case OBJECT_AMOP:
+		case OBJECT_AMPROC:
+		{
+			objnode = (Node *) list_make2(name, args);
+			break;
+		}
+
+		case OBJECT_FUNCTION:
+		case OBJECT_PROCEDURE:
+		case OBJECT_ROUTINE:
+		case OBJECT_AGGREGATE:
+		case OBJECT_OPERATOR:
+		{
+			ObjectWithArgs *owa = makeNode(ObjectWithArgs);
+
+			owa->objname = name;
+			owa->objargs = args;
+			objnode = (Node *) owa;
+			break;
+		}
+
+		case OBJECT_LARGEOBJECT:
+		{
+			/* already handled above */
+			break;
+		}
+
+			/* no default, to let compiler warn about missing case */
+	}
+
+	if (objnode == NULL)
+	{
+		elog(ERROR, "unrecognized object type: %d", type);
+	}
+
+	ObjectAddress addr = get_object_address(type, objnode,
+											&relation, AccessShareLock, false);
+
+	/* CITUS CODE BEGIN */
+
+	if (accessCheck)
+	{
+		Oid userId = GetUserId();
+		AclMode aclMaskResult = 0;
+		Oid idToCheck = InvalidOid;
+
+		switch (type)
+		{
+			case OBJECT_ACCESS_METHOD:
+			{
+				if (IsObjectAddressOwnedByExtension(&addr, NULL))
+				{
+					elog(ERROR,
+						 "Current user does not have required access privileges on access method %d with type %d",
+						 addr.objectId, type);
+				}
+				aclMaskResult = 1;
+				break;
+			}
+
+			case OBJECT_SCHEMA:
+			{
+				idToCheck = addr.objectId;
+				aclMaskResult = pg_namespace_aclmask(idToCheck, userId, ACL_USAGE |
+													 ACL_CONNECT | ACL_INSERT |
+													 ACL_SELECT, ACLMASK_ANY);
+				break;
+			}
+
+			case OBJECT_FUNCTION:
+			case OBJECT_PROCEDURE:
+			case OBJECT_AGGREGATE:
+			{
+				idToCheck = addr.objectId;
+				aclMaskResult = pg_proc_aclmask(idToCheck, userId, ACL_EXECUTE,
+												ACLMASK_ANY);
+				break;
+			}
+
+			case OBJECT_DATABASE:
+			{
+				idToCheck = addr.objectId;
+				aclMaskResult = pg_database_aclmask(idToCheck, userId, ACL_CONNECT,
+													ACLMASK_ANY);
+				break;
+			}
+
+			case OBJECT_ROLE:
+			{
+				if (!(addr.objectId == CitusExtensionOwner()))
+				{
+					elog(ERROR,
+						 "Current user does not have required access privileges on role %d with type %d",
+						 addr.objectId, type);
+				}
+				aclMaskResult = 1;
+				break;
+			}
+
+			case OBJECT_TYPE:
+			{
+				idToCheck = addr.objectId;
+				aclMaskResult = pg_type_aclmask(idToCheck, userId, ACL_USAGE,
+												ACLMASK_ANY);
+				break;
+			}
+
+			case OBJECT_SEQUENCE:
+			case OBJECT_TABLE:
+			{
+				/* function errors out if user is not the owner of the object */
+				check_object_ownership(userId, type, addr, objnode, relation);
+				aclMaskResult = 1;
+				break;
+			}
+
+			case OBJECT_COLLATION:
+			case OBJECT_EXTENSION:
+			{
+				aclMaskResult = 1;
```

Comment:
Well, as of PG 14, only superusers can create extensions. Should we add superuser checks here?

Also, for collations, why do we allow by default ?